### PR TITLE
Feature implementation: spotlight()

### PR DESCRIPTION
**Pull request by sanketsingh24**
Rebased on top of [`feature-specularlight`](https://github.com/processing/p5.js/pull/3843), so that needs to be merged first.

fixes- #3894 

See manual test `/webgl/lights/spotlight/`.

Work done- 
- [x] Implemented spotlight uniform in `lighting.glsl`.
- [x] Implemented `spotLight()` in /src/webgl/light.js
- [x] Added a manual-test-example in `/webgl/lights/spotLight/`
- [x] Unit tests
- [x] Documentation
- [x] Example

To do-

_More to follow..._

### CODE REVIEW

**Comment by aferriss** [comment]
I think that this PR is in a pretty good state (once the conflicts from specularLight getting merged are resolved). I wanted to again call out to see if @stalgiag @Spongman  or anyone else had feedback on this feature or the PR before it gets merged.

File: test/manual-test-examples/webgl/lights/spotLight/index.html [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  <meta http-equiv="X-UA-Compatible" content="IE=edge">
+  <title></title>
+  <link rel="stylesheet" href="../../../styles.css">
+  <script language="javascript" type="text/javascript" src="../../../../../lib/p5.js"></script>
+  <script language="javascript" type="text/javascript" src="sketch.js"></script>
+  <script src="../../stats.js"></script>
+</head>
+
+<body>
+  <header>
```
Comment (stalgiag) [inline test/manual-test-examples/webgl/lights/spotLight/index.html]: Small note, can you remove this? Not applicable here as far as I can tell

File: src/webgl/light.js [hunk 2]
```diff
@@ -492,4 +492,344 @@ p5.prototype.lightFalloff = function(
...
+  angle,
+  concentration
+) {
+  this._assert3d('spotLight');
+  p5._validateParameters('spotLight', arguments);
+
+  let color, position, direction;
+  const length = arguments.length;
+
+  switch (length) {
```
Comment (stalgiag) [inline src/webgl/light.js]: Good job juggling so many different parameters!

File: src/webgl/light.js [hunk 3]
```diff
@@ -492,4 +492,344 @@ p5.prototype.lightFalloff = function(
...
+ * @param  {Number}    [angle]  optional parameter for angle. Defaults to PI/3
+ * @param  {Number}    [conc]   optional parameter for concentration. Defaults to 100
+ * @chainable
+ *
+ * @example
+ * <div>
+ * <code>
+ * function setup() {
+ *   createCanvas(100, 100, WEBGL);
+ *   setAttributes('perPixelLighting', true);
```
Comment (stalgiag) [inline src/webgl/light.js]: Part of me feels like we should set 'perPixelLighting' to true by default. These lines can feel unnecessarily complicating for the lighting references.

Comment (aferriss) [reply src/webgl/light.js]: I agree, and I think the per-vertex lighting looks pretty ugly. I also feel like the setAttributes function is a little arcane.

Comment (stalgiag) [reply src/webgl/light.js]: Yeah I am not sure if there are performance hits with per pixel. The `setAttributes` function is definitely intended to only be used by people that want a non-traditional webgl setup for whatever reason. The most commonly used attribute modification would theoretically be 'antialias' which would likely be changed by `smooth()` for most users.

I'll break this out into an issue.

**Pull request merged**